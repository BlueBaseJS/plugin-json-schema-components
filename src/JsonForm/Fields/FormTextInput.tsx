import { TextInput, TextInputProps } from '@bluebase/components';
import { getComponent } from '@bluebase/core';
import { useField } from 'formik';
import React from 'react';

import { removeUndefinedProperties } from '../../helpers';
import { BaseFormFieldProps } from '../BaseFormField';

const BaseFormField = getComponent<BaseFormFieldProps>('BaseFormField');

export type FormTextInputProps<T = {}> = TextInputProps &
	BaseFormFieldProps &
	T & {
		children?: React.ReactNode;
		validate?: (value: any) => string | Promise<void> | undefined;
		name: string;
		type?: string;
		value?: any;
		innerRef?: (instance: any) => void;
	};

export const validateFormTextInput = (props: FormTextInputProps) => (value: string) => {
	const { required, type } = props;

	let error;

	// If field is required
	if (required === true && !value) {
		error = 'This field is required';
	}

	// If field 'email'
	else if (type === 'email' && !/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(value)) {
		error = 'Please enter a valid email address';
	}

	return error;
};

export const FormTextInput = (props: FormTextInputProps) => {
	const [field, meta, helpers] = useField({
		...(props as any),
		validate: props.validate || validateFormTextInput(props),
	});

	const inputProps = {
		...field,
		onChange: undefined,
		...props,
		error: !!meta.error || props.error,
		helperText: meta.error || props.helperText,
		onChangeText: helpers.setValue,

		// Fix for https://github.com/BlueBaseJS/plugin-json-schema-components/issues/32
		// Needs to be investigated why this is happening
		onBlur: undefined,
	};

	return <BaseFormField {...removeUndefinedProperties(inputProps)} />;
};

FormTextInput.defaultProps = {
	MainComponent: TextInput,
};

FormTextInput.displayName = 'FormTextInput';
