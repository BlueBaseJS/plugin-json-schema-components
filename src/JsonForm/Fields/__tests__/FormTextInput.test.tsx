import { BlueBaseApp } from '@bluebase/core';
import { FormTextInput } from '../FormTextInput';
import { Formik } from 'formik';
import Plugin from '../../../index';
import React from 'react';
import { mount } from 'enzyme';
import { waitForElement } from 'enzyme-async-helpers';

const fieldProps = [{
	label: 'Username',
	name: 'username',
	required: true,
	type: 'text',
}, {
	label: 'Email',
	name: 'email',
	required: true,
	type: 'email',
}];

const initialValues = {
	username: 'foo',
};


describe('FormTextInput', () => {


	it('should render all form fields in the schema', async () => {

		const onSubmit = jest.fn();

		const component = mount(
			<BlueBaseApp plugins={[Plugin]}>
				<Formik initialValues={initialValues} onSubmit={onSubmit}>
					<FormTextInput {...fieldProps[0]} />
				</Formik>
			</BlueBaseApp>
		);

		await waitForElement(component as any, FormTextInput);

		// expect(component).toMatchSnapshot();

		// Check fields
		expect(component.find('TextInput').last().prop('name')).toBe('username');
		expect(component.find('TextInput').last().prop('value')).toBe('foo');
		expect(component.find('TextInput').last().prop('type')).toBe('text');
	});


	it('should validate required fields', async () => {

		const onSubmit = jest.fn();

		const component = mount(
			<BlueBaseApp plugins={[Plugin]}>
				<Formik initialValues={initialValues} onSubmit={onSubmit}>
					<FormTextInput {...fieldProps[0]} />
				</Formik>
			</BlueBaseApp>
		);

		await waitForElement(component as any, FormTextInput);

		// expect(component).toMatchSnapshot();

		const validate: any = component.find('FieldInner[name="username"]').last().prop('validate');

		// Check fields
		expect(validate('')).toBe('This field is required');
		expect(validate('foo')).toBeUndefined();
	});


	it('should validate email fields', async () => {

		const onSubmit = jest.fn();

		const component = mount(
			<BlueBaseApp plugins={[Plugin]}>
				<Formik initialValues={initialValues} onSubmit={onSubmit}>
					<FormTextInput {...fieldProps[1]} />
				</Formik>
			</BlueBaseApp>
		);

		await waitForElement(component as any, FormTextInput);

		// expect(component).toMatchSnapshot();

		const validate: any = component.find('FieldInner[name="email"]').last().prop('validate');

		// Check fields
		expect(validate('foo bar')).toBe('Please enter a valid email address');
		expect(validate('foo@bar.com')).toBeUndefined();
	});
});
