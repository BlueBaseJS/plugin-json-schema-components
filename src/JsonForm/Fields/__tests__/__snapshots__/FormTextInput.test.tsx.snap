// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormTextInput should render all form fields in the schema 1`] = `
<Formik
  initialValues={
    Object {
      "username": "foo",
    }
  }
  onSubmit={[MockFunction]}
>
  <FormTextInput
    MainComponent={[Function]}
    helperText="This is a helper text"
    label="Username"
    name="username"
    required={true}
    type="text"
  >
    <BaseFormField
      MainComponent={[Function]}
      helperText="This is a helper text"
      label="Username"
      name="username"
      onChangeText={[Function]}
      required={true}
      type="text"
      value="foo"
    >
      <BaseFormField
        MainComponent={[Function]}
        helperText="This is a helper text"
        label="Username"
        name="username"
        onChangeText={[Function]}
        required={true}
        type="text"
        value="foo"
      >
        <Component
          MainComponent={[Function]}
          helperText="This is a helper text"
          label="Username"
          name="username"
          onChangeText={[Function]}
          required={true}
          type="text"
          value="foo"
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "flex-start",
                "paddingHorizontal": 16,
                "paddingVertical": 16,
              }
            }
            testID="base-form-field"
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "flex-start",
                  "paddingHorizontal": 16,
                  "paddingVertical": 16,
                }
              }
              testID="base-form-field"
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "flex-start",
                    "paddingHorizontal": 16,
                    "paddingVertical": 16,
                  }
                }
                testID="base-form-field"
              >
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "flex-start",
                      "paddingHorizontal": 16,
                      "paddingVertical": 16,
                    }
                  }
                  testID="base-form-field"
                >
                  <View
                    style={
                      Object {
                        "flex": 1,
                      }
                    }
                    testID="base-form-field-main"
                  >
                    <View
                      style={
                        Object {
                          "flex": 1,
                        }
                      }
                      testID="base-form-field-main"
                    >
                      <View
                        style={
                          Object {
                            "flex": 1,
                          }
                        }
                        testID="base-form-field-main"
                      >
                        <View
                          style={
                            Object {
                              "flex": 1,
                            }
                          }
                          testID="base-form-field-main"
                        >
                          <TextInput
                            helperText="This is a helper text"
                            label="Username"
                            name="username"
                            onChangeText={[Function]}
                            required={true}
                            type="text"
                            value="foo"
                          >
                            <TextInput
                              helperText="This is a helper text"
                              label="Username"
                              name="username"
                              onChangeText={[Function]}
                              required={true}
                              type="text"
                              value="foo"
                            >
                              <Component
                                allowFontScaling={true}
                                helperText="This is a helper text"
                                label="Username"
                                name="username"
                                onChangeText={[Function]}
                                rejectResponderTermination={true}
                                required={true}
                                type="text"
                                underlineColorAndroid="transparent"
                                value="foo"
                              >
                                <TextInput
                                  allowFontScaling={true}
                                  helperText="This is a helper text"
                                  label="Username"
                                  name="username"
                                  onChangeText={[Function]}
                                  rejectResponderTermination={true}
                                  required={true}
                                  type="text"
                                  underlineColorAndroid="transparent"
                                  value="foo"
                                />
                              </Component>
                            </TextInput>
                          </TextInput>
                        </View>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </Component>
      </BaseFormField>
    </BaseFormField>
  </FormTextInput>
</Formik>
`;
